build-system/app-index/test/test-amphtml-helpers.js

```diff
@@ -0,0 +1,411 @@
+
+      it('adds mixed', () => {
+
+        const expectedExtensions = [
+          'amp-foo',
+          'amp-bar',
+          'amp-foo-bar-baz',
+          'amp-bind',
+          'amp-form',
+        ];
+
+        const expectedTemplates = ['amp-mustache'];
+
+        const rawStr = html`
+          <html>
+            <head></head>
+            <body>
+              <amp-foo foo=bar></amp-foo>
+              <amp-foo foo=bar></amp-foo>
+              <amp-foo foo=bar></amp-foo>
+              <div>
+                <amp-bar></amp-bar>
+                <div>
+                  <amp-foo-bar-baz many=1 attributes=2>
+                    Text content
+                  </amp-foo-bar-baz>
+                </div>
+                <input>
+                <amp-state id=myState></amp-state>
+                <template type="amp-mustache"></template>
+              </div>
+            </body>
+          </html>`;
+
+        const {document} =
+            (new JSDOM(addRequiredExtensionsToHead(rawStr))).window;
+
+        const scripts =
+            Array.from(document.head.getElementsByTagName('script'));
+
+        expect(scripts.length)
+            .to.equal(expectedExtensions.length + expectedTemplates.length);
+
+        scripts.forEach(script => {
+          expect(script.getAttribute('src')).to.be.ok;
+          expect(script.getAttribute('async')).to.equal('');
+        });
+
+        expectedExtensions.forEach(expectedScript => {
+          expect(scripts.some(s => {
+            if (s.getAttribute('custom-element') == expectedScript) {
+              expect(s.getAttribute('custom-template')).to.be.null;
+              return true;
+            }
+            return false;
+          })).to.be.true;
+        });
+
+        expectedTemplates.forEach(expectedScript => {
+          expect(scripts.some(s => {
+            if (s.getAttribute('custom-template') == expectedScript) {
+              expect(s.getAttribute('custom-extension')).to.be.null;
+              return true;
+            }
+            return false;
+          })).to.be.true;
+        });
+      });
+
+      it('adds extensions', () => {
+
+        const expected = ['amp-foo', 'amp-bar', 'amp-foo-bar-baz'];
+
+        const rawStr = html`
+          <html>
+            <head></head>
+            <body>
+              <amp-foo foo=bar></amp-foo>
+              <amp-foo foo=bar></amp-foo>
+              <amp-foo foo=bar></amp-foo>
+              <div>
+                <amp-bar></amp-bar>
+                <div>
+                  <amp-foo-bar-baz many=1 attributes=2>
+                    Text content
+                  </amp-foo-bar-baz>
+                </div>
+              </div>
+            </body>
+          </html>`;
+
+        const {document} =
+            (new JSDOM(addRequiredExtensionsToHead(rawStr))).window;
+
+        const scripts =
+            Array.from(document.head.getElementsByTagName('script'));
+
+        expect(scripts.length).to.equal(expected.length);
```

Comment:
expect(scripts).to.have.length(expected.length);

that way, if it has the wrong length, it should give a more informative message than something like "expected 5 to equal 4"